File: packages/sdk-python/veto/integrations/langchain/tool_node.py
"""
LangGraph ToolNode wrapper for Veto.

Validates every tool call from the model's response before delegating to
the actual ``ToolNode`` for execution. Denied calls return a ``ToolMessage``
with the denial reason instead of executing the tool.

Usage::

    from langgraph.prebuilt import ToolNode
    from veto import Veto, VetoOptions
    from veto.integrations.langchain import create_veto_tool_node

    veto = await Veto.init(VetoOptions(api_key="your-key"))
    tool_node = ToolNode([search_tool, email_tool])
    veto_tool_node = create_veto_tool_node(veto, tool_node)

    # Use veto_tool_node in your LangGraph graph
    graph = StateGraph(MessagesState)
    graph.add_node("tools", veto_tool_node)
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any, Callable, Optional

if TYPE_CHECKING:
    from veto.core.veto import Veto

from veto.types.tool import ToolCall
from veto.utils.id import generate_tool_call_id

logger = logging.getLogger("veto.integrations.langchain")


def create_veto_tool_node(
    veto: "Veto",
    tool_node: Any,
    *,
    on_allow: Optional[Callable[..., Any]] = None,
    on_deny: Optional[Callable[..., Any]] = None,
) -> Callable[..., Any]:
    """
    Wrap a LangGraph ``ToolNode`` with Veto validation.

    Returns an async function with the same signature as a LangGraph node
    that validates each tool call in the last ``AIMessage`` before
    delegating to the real ``ToolNode``.

    Args:
        veto: An initialized ``Veto`` instance.
        tool_node: A LangGraph ``ToolNode`` instance.
        on_allow: Optional callback ``(tool_name, args) -> None``.
        on_deny: Optional callback ``(tool_name, args, reason) -> None``.

    Returns:
        An async function usable as a LangGraph node.
    """

    async def validated_tool_node(state: dict[str, Any]) -> dict[str, Any]:
        messages = state.get("messages", [])
        if not messages:
            return await tool_node.ainvoke(state)

        last_message = messages[-1]
        tool_calls = getattr(last_message, "tool_calls", None) or []

        if not tool_calls:
            return await tool_node.ainvoke(state)

        for tc in tool_calls:
            if isinstance(tc, dict):
                name = tc.get("name", "")
                args = tc.get("args", {})
                call_id = tc.get("id") or generate_tool_call_id()
            else:
                name = getattr(tc, "name", "")
                args = getattr(tc, "args", {})
                call_id = getattr(tc, "id", None) or generate_tool_call_id()

            result = await veto._validate_tool_call(
                ToolCall(
                    id=call_id,
                    name=name,
                    arguments=args if isinstance(args, dict) else {},
                )
            )

            if not result.allowed:
                reason = result.validation_result.reason or "Policy violation"
                logger.info("BLOCKED %s: %s", name, reason)

                if on_deny is not None:
                    await on_deny(name, args, reason)

                try:
                    from langchain_core.messages import ToolMessage
                    return {
                        "messages": [
                            ToolMessage(
                                content=f"Tool call denied by Veto: {reason}",
                                tool_call_id=t.get("id", call_id) if isinstance(t, dict) else getattr(t, "id", call_id),
                            )
                            for t in tool_calls
                        ]
                    }
                except ImportError:
                    return {
                        "messages": [
                            {
                                "content": f"Tool call denied by Veto: {reason}",
                                "tool_call_id": t.get("id", call_id) if isinstance(t, dict) else getattr(t, "id", call_id),
                            }
                            for t in tool_calls
                        ]
                    }

            logger.info("ALLOWED %s", name)
            if on_allow is not None:
                await on_allow(name, args)

        return await tool_node.ainvoke(state)

    return validated_tool_node

File: packages/sdk/src/integrations/langchain/tool-node.ts
import type { Veto } from '../../core/veto.js';
import { generateToolCallId } from '../../utils/id.js';

export interface VetoToolNodeOptions {
  /** Called when a tool call is allowed. */
  onAllow?: (toolName: string, args: Record<string, unknown>) => void | Promise<void>;
  /** Called when a tool call is denied. */
  onDeny?: (toolName: string, args: Record<string, unknown>, reason: string) => void | Promise<void>;
}

async function loadOptionalModule<T>(moduleName: string): Promise<T> {
  return await import(moduleName) as T;
}

/**
 * Create a LangGraph-compatible node function that validates tool calls
 * through Veto before delegating to a real ToolNode.
 *
 * This is for users building raw LangGraph graphs who want per-node
 * validation without using `createAgent` + middleware.
 *
 * @example
 * ```ts
 * import { StateGraph, MessagesAnnotation, START, END } from '@langchain/langgraph';
 * import { ToolNode, toolsCondition } from '@langchain/langgraph/prebuilt';
 * import { Veto } from 'veto-sdk';
 * import { createVetoToolNode } from 'veto-sdk/integrations/langchain';
 *
 * const veto = await Veto.init();
 * const toolNode = new ToolNode([searchTool, emailTool]);
 * const vetoToolNode = createVetoToolNode(veto, toolNode);
 *
 * const graph = new StateGraph(MessagesAnnotation)
 *   .addNode('agent', callModel)
 *   .addNode('tools', vetoToolNode)
 *   .addEdge(START, 'agent')
 *   .addConditionalEdges('agent', toolsCondition, ['tools', END])
 *   .addEdge('tools', 'agent')
 *   .compile();
 * ```
 */
export function createVetoToolNode(
  veto: Veto,
  toolNode: { invoke: (state: any) => Promise<any> },
  options?: VetoToolNodeOptions,
): (state: any) => Promise<any> {
  const onAllow = options?.onAllow;
  const onDeny = options?.onDeny;

  return async (state: { messages: any[] }) => {
    const lastMessage = state.messages[state.messages.length - 1];

    const toolCalls: Array<{ name: string; args: Record<string, unknown>; id?: string }> =
      lastMessage?.tool_calls ?? [];

    if (toolCalls.length === 0) {
      return toolNode.invoke(state);
    }

    // Validate ALL tool calls before deciding.
    // Track denied by index to avoid ID mismatch when tool calls lack an explicit id.
    const denied: Array<{ callId: string; reason: string; index: number }> = [];
    const deniedIndices = new Set<number>();

    for (let i = 0; i < toolCalls.length; i++) {
      const tc = toolCalls[i];
      const callId = tc.id ?? generateToolCallId();

      const result = await veto.validateToolCall({
        id: callId,
        name: tc.name,
        arguments: tc.args ?? {},
      });

      if (!result.allowed) {
        const reason = result.validationResult?.reason ?? 'Policy violation';
        denied.push({ callId, reason, index: i });
        deniedIndices.add(i);
        if (onDeny) await onDeny(tc.name, tc.args, reason);
      } else {
        if (onAllow) await onAllow(tc.name, tc.args);
      }
    }

    if (denied.length === 0) {
      return toolNode.invoke(state);
    }

    // Build denial messages
    let ToolMessage: any;
    try {
      const mod = await loadOptionalModule<{ ToolMessage: new (body: Record<string, unknown>) => any }>(
        '@langchain/core/messages',
      );
      ToolMessage = mod.ToolMessage;
    } catch {
      ToolMessage = null;
    }

    const denialMessages = denied.map(d =>
      ToolMessage
        ? new ToolMessage({ content: `Tool call denied by Veto: ${d.reason}`, tool_call_id: d.callId })
        : { content: `Tool call denied by Veto: ${d.reason}`, tool_call_id: d.callId },
    );

    // If ALL calls denied, return denials only
    if (denied.length === toolCalls.length) {
      return { messages: denialMessages };
    }

    // Partial denial — execute allowed calls, merge with denial messages
    const allowedCalls = toolCalls.filter((_: any, idx: number) => !deniedIndices.has(idx));
    const modifiedMessages = [...state.messages];
    const lastMsg = modifiedMessages[modifiedMessages.length - 1];
    modifiedMessages[modifiedMessages.length - 1] = { ...lastMsg, tool_calls: allowedCalls };

    const allowedResult = await toolNode.invoke({ ...state, messages: modifiedMessages });
    return {
      messages: [...denialMessages, ...(allowedResult?.messages ?? [])],
    };
  };
}
